sipyco/sync_struct.py

```diff
@@ -319,11 +361,19 @@ def __init__(self, notifiers):
             try:
                 while True:
                     line = await queue.get()
+                    if line is None:
+                        # Notifier removed, close connection.
+                        return
                     writer.write(line)
                     # raise exception on connection error
                     await writer.drain()
             finally:
-                self._recipients[notifier_name].remove(queue)
+                try:
+                    self._recipients[notifier_name].remove(queue)
+                except KeyError:
+                    # Notifier might have been removed using remove_notifier
+                    # already.
```

Comment:
Won't this break if you remove a notifier, then add it again immediately with the same name?

Reply:
Hm, what would we expect for existing connections in that scenario? (This just deals with the one connection queue, doesn't it?) I think closing them would be fine, although, granted, we are never re-adding listeners in our use cases.

Reply:
What is the use scenario exactly?

Reply:
@dnadlinger will be better-placed to comment than me, but my (limited) understanding is that this is an essential feature for ndscan.

One of the really nice things about ndscan is that the applets know a lot about the scan. So, e.g. one has a single applet which can handle scanning an arbitrary number of dimensions, can set up axes correctly, etc without manual intervention (which is not possible in artiq). There are also really nice features like being able to update a dataset from scan results by right clicking in an applet.

The scans needs to be able to run in parallel, interrupting each other.

The way the scans are synchronised with the applets is using a separate notifier for each RID (namespaced datasets). Then using datasets to broadcast metadata. https://github.com/OxfordIonTrapGroup/ndscan/blob/7dd4a9de5255bf5a92a38750e020d0eaa7b17385/ndscan/experiment/entry_point.py#L445-L472